log/gps/gps_mgl_enigma_waypoint.ksy

```diff
@@ -0,0 +1,148 @@
+    MGL Avionics places this document and the data format it describes in the public domain in order to foster the use of a common, compact waypoint and route file format that is particularly optimized for use with resource limited systems and/or systems that require fast access to individual items in the waypoint files in order to increase overall systems performance.
+    In order to allow recognition of the format any party adopting this format agrees to refer to this format as the “Enigma waypoint format”. All data using this format should be considered to be in the public domain and the implementer will make reasonable effort to allow any interested party to use the data.
+    Exceptions:
+      MGL Avionics reserves copyright and denies use of this data format for any military or related purpose.
+
+doc: |
+  Waypoint format used by Enigma series of EFIS instruments from MGL Avionics.
+  It is common to create an index file containing a simple list of indexes, one to each entry in the waypoint file, sorted by a particular criteria (for example distance of the waypoint to a given location), rather than sorting the actual waypoint list itself. This is a method commonly employed by database applications.
+
+doc-ref:
+  - https://mglavionics.co.za/Docs/Enigma%20Waypoint%20format.pdf
+
+seq:
+  - id: waypoints
+    type: waypoint
+    repeat: eos
+    doc: There is no demand that waypoints are stored in any particular order in the waypoint file. They might be sorted by short name or waypoint type but this is of little consequence. The order of waypoints in the file matters only for route files. Route waypoints are stored in the file in order of their appearance in the route. A route should have at least one waypoint.
+
+types:
+  waypoint:
+    seq:
+      - id: lat
+        type: sph_coord
+        doc: Positive values are in Nothern Hemisphere
+      - id: lon
+        type: sph_coord
+        doc: Positive values are in Eastern Hemisphere
+      - id: data_raw
+        type: u4
+      - id: type
+        type: u1
+        enum: type
+      - id: name_short
+        type: dc_str(6)
+        doc: This entry is used as “key” and in aviation terms would contain the short identifier of an airfield or navaid beacon. This key is used to lookup related data in other databases, for example an airport data base containing details of an airport such as frequencies and runway headings.
+      - id: name_long
+        type: dc_str(27)
+        doc: |
+          The long name contains a human readable description of the waypoint.
+          For example: “Los Angeles International”. The long name is arbitrary and not used as search key for any purpose.
+
+    instances:
+      is_data_frequency:
+        value: 9 <= type.to_i and type.to_i <= 25
+      altitude:
+        value: "( ( data_raw < 0x80000000 ) ? data_raw : ( ( data_raw & 0x7FFFFFFF ) - 0x80000000 ) )"
```

Comment:
The problem is the field on which the interpretation of this field depends (`type`) follows this field. It can be done using a `pos` `instance`, but it would be heavier.

Reply:
@KOLANICH:

> The problem is the field on which the interpretation of this field depends (`is_data_frequency` is derived from `type`) follows this field. It can be done using a `pos` `instance`, but it would be heavier.

```ksy
    seq:
      # ...
      - id: data_slot
        size: 4
        type: slot
      - id: type
        type: u1
        enum: type
    instances:
      altitude:
        io: data_slot._io
        pos: 0
        type: s4
        if: not is_data_frequency
        doc: in feet
      frequency:
        io: data_slot._io
        pos: 0
        type: u4
        if: is_data_frequency
        doc: in kHz
    types:
      slot: {}
```

Reply:
Yeah, a KaitaiStream object + a buffer object to store 4 bytes (that would likely consume not less than 8 bytes of overhead) + some reuses of tuat stream object .... Feels like much (and unnecessarily) heavier than just manually converting unsigned into signed.

Reply:
> Yeah, a KaitaiStream object + a buffer object to store 4 bytes (that would likely consume not less than 8 bytes of overhead) + some reuses of tuat stream object

The overhead will be insignificant when we implement data views / substreams (https://github.com/kaitai-io/kaitai_struct/issues/44).